File: hermes/src/main/java/org/eclipse/paho/android/service/MqttConnection.java
/*******************************************************************************
 * Copyright (c) 1999, 2016 IBM Corp.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * and Eclipse Distribution License v1.0 which accompany this distribution. 
 *
 * The Eclipse Public License is available at 
 *    http://www.eclipse.org/legal/epl-v10.html
 * and the Eclipse Distribution License is available at 
 *   http://www.eclipse.org/org/documents/edl-v10.php.
 */
package org.eclipse.paho.android.service;

import android.accounts.NetworkErrorException;
import android.content.Context;
import android.os.PowerManager.WakeLock;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.eclipse.paho.android.service.bean.ClientInfoBean;
import org.eclipse.paho.android.service.contract.DisconnectMqttEventActionListener;
import org.eclipse.paho.android.service.contract.PublishMqttEventActionListener;
import org.eclipse.paho.android.service.contract.SubscribeMqttEventActionListener;
import org.eclipse.paho.android.service.contract.UnsubscribeMqttEventActionListener;
import org.eclipse.paho.android.service.db.DbStoredData;
import org.eclipse.paho.android.service.db.MessageStore;
import org.eclipse.paho.android.service.event.MqttAction;
import org.eclipse.paho.android.service.event.MqttEvent;
import org.eclipse.paho.android.service.sender.AlarmPingSender;
import org.eclipse.paho.client.mqttv3.DisconnectedBufferOptions;
import org.eclipse.paho.client.mqttv3.IMqttActionListener;
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.IMqttMessageListener;
import org.eclipse.paho.client.mqttv3.IMqttToken;
import org.eclipse.paho.client.mqttv3.MqttAsyncClient;
import org.eclipse.paho.client.mqttv3.MqttCallbackExtended;
import org.eclipse.paho.client.mqttv3.MqttClientPersistence;
import org.eclipse.paho.client.mqttv3.MqttMessage;
import org.greenrobot.eventbus.EventBus;

import java.util.ArrayList;
import java.util.Arrays;

public class MqttConnection implements MqttCallbackExtended {

	private static final String TAG = "MqttConnection";

	private final Context mContext;

	/**
	 * 客户端
	 */
	@Nullable
	private  MqttAsyncClient mClient = null;

	/**
	 * ping数据包发送器
	 */
	private final AlarmPingSender mAlarmPingSender;

	/**
	 * 是否连接中
	 */
	private volatile boolean isConnecting = false;

	/** 唤醒锁 */
	@NonNull
	private final WakeLock mWakeLock;

	/** 数据库操作接口 */
	@NonNull
	public final MessageStore mMessageStore;

	private final ClientInfoBean mBean;

	private DisconnectedBufferOptions bufferOpts = null;

	public MqttConnection(Context context, ClientInfoBean bean, @NonNull MqttClientPersistence persistence, @NonNull MessageStore messageStore) {
		mContext = context;
		mBean = bean;
		mMessageStore = messageStore;
		mWakeLock = MqttUtils.getWakeLock(context, getClass().getName() + " " + bean.getClientId() + " " + "on host " + bean.getServerURI());
		mAlarmPingSender = new AlarmPingSender(mContext);
		try {
			mClient = new MqttAsyncClient(bean.getServerURI(), bean.getClientId(), persistence, mAlarmPingSender);
			mClient.setCallback(this);
		}catch (Exception e){
			e.printStackTrace();
			EventBus.getDefault().post(MqttEvent.createFail(mBean.getClientKey(), MqttAction.ACTION_CREATE_CLIENT, "create MqttAsyncClient fail",e));
		}
	}

	/** 连接服务器 */
	public void connect() {
		if (mClient == null) {
			EventBus.getDefault().post(MqttEvent.createConnectFail(mBean.getClientKey(), "client is null", new NullPointerException()));
			return;
		}
		Log.d(TAG, "start connect {" + mBean.getClientKey() + "}");
		if (mBean.getConnectOptions().isCleanSession()) { // 如果是新的会话则清空数据库缓存
			mMessageStore.clearAllMessages(mBean.getClientKey());
		}

		if (isConnecting) {
			Log.d(TAG, "now connecting");
			return;
		}
		if (mClient.isConnected()) {
			Log.d(TAG, "client is connected");
			connectSuccess();
			return;
		}
		Log.d(TAG, "start connect!");
		isConnecting = true;
		try {
			mClient.connect(mBean.getConnectOptions(), mConnectActionListener);

		} catch (Exception e) {
			e.printStackTrace();
			isConnecting = false;
			EventBus.getDefault().post(MqttEvent.createConnectFail(mBean.getClientKey(), "connect fail", e));
		}
	}

	private final IMqttActionListener mConnectActionListener = new IMqttActionListener() {
		@Override
		public void onSuccess(IMqttToken asyncActionToken) {
			Log.d(TAG, "connect success!");
			connectSuccess();
		}

		@Override
		public void onFailure(IMqttToken asyncActionToken, Throwable exception) {
			Log.e(TAG, "connect fail, call connect to reconnect.reason:" + exception.getMessage());
			connectFail("connect failure", exception);
		}
	};

	/** 连接成功 */
	private void connectSuccess() {
		MqttUtils.acquireWakeLock(mWakeLock);
		EventBus.getDefault().post(MqttEvent.createConnectSuccess(mBean.getClientKey()));
		deliverBacklog();
		isConnecting = false;
		MqttUtils.releaseWakeLock(mWakeLock);
	}

	/**
	 * 将数据库缓存的到达消息全部返回给用户
	 */
	private void deliverBacklog() {
		ArrayList<DbStoredData> backlog = mMessageStore.getAllMessages(mBean.getClientKey());
		for (DbStoredData data : backlog) {
			EventBus.getDefault().post(MqttEvent.createMsgArrived(mBean.getClientKey(), data));
		}
	}

	/** 连接失败 */
	private void connectFail(String errorMsg, Throwable e){
		MqttUtils.acquireWakeLock(mWakeLock);
		isConnecting = false;
		EventBus.getDefault().post(MqttEvent.createConnectFail(mBean.getClientKey(), errorMsg, e));
		MqttUtils.releaseWakeLock(mWakeLock);
	}

	/** 关闭客户端 */
	void close() {
		Log.d(TAG, "close client");
		try {
			if (mClient != null) {
				mClient.close();
			}
			EventBus.getDefault().post(MqttEvent.createSuccess(mBean.getClientKey(), MqttAction.ACTION_CLOSE));
		} catch (Exception e) {
			e.printStackTrace();
			EventBus.getDefault().post(MqttEvent.createFail(mBean.getClientKey(), MqttAction.ACTION_CLOSE, "close client fail", e));
		}
	}

	/**
	 * 断开连接
	 */
	public void disconnect() {
		disconnect(-1);
	}

	/**
	 * 断开连接
	 * @param quiesceTimeout 在断开连接之前允许完成现有工作的时间（以毫秒为单位）。 值为零或更低意味着客户端不会停顿。
	 */
	public void disconnect(long quiesceTimeout) {
		Log.d(TAG, "mqtt disconnect {" + mBean.getClientKey() + "}");
		if (mClient == null) {
			EventBus.getDefault().post(MqttEvent.createFail(mBean.getClientKey(), MqttAction.ACTION_DISCONNECT, "client is null", new NullPointerException()));
			return;
		}
		if (!mClient.isConnected()) {
			EventBus.getDefault().post(MqttEvent.createSuccess(mBean.getClientKey(), MqttAction.ACTION_DISCONNECT));
			return;
		}
		Log.d(TAG, "start disconnect!");
		IMqttActionListener listener = new DisconnectMqttEventActionListener(mBean.getClientKey(), "disconnect failure");
		try {
			if (quiesceTimeout > 0) {
				mClient.disconnect(quiesceTimeout, null, listener);
			} else {
				mClient.disconnect(null, listener);
			}
		} catch (Exception e) {
			e.printStackTrace();
			EventBus.getDefault().post(MqttEvent.createFail(mBean.getClientKey(), MqttAction.ACTION_DISCONNECT, "disconnect fail", e));
		}
		if (mBean.getConnectOptions().isCleanSession()) {
			mMessageStore.clearAllMessages(mBean.getClientKey());
		}
		MqttUtils.releaseWakeLock(mWakeLock);
	}

	/** 是否已连接 */
	public boolean isConnected() {
		return mClient != null && mClient.isConnected();
	}

	/**
	 * 向主题发送消息
	 * @param topic 主题名称
	 * @param message 消息对象
	 */
	public IMqttDeliveryToken publish(String topic, MqttMessage message) {
		if (mClient == null) {
			EventBus.getDefault().post(MqttEvent.createPublishFail(mBean.getClientKey(), topic, "client is null", new NullPointerException()));
			return null;
		}
		IMqttActionListener listener =  new PublishMqttEventActionListener(mBean.getClientKey(), "publish failure", topic, message);
		try {
			if (mClient.isConnected()) {
				return mClient.publish(topic, message, null, listener);
			}
			if (this.bufferOpts != null && this.bufferOpts.isBufferEnabled()) {
				// 虽然客户端未连接，但是缓冲已启用，也允许发送消息
				return mClient.publish(topic, message, null, listener);
			}
			EventBus.getDefault().post(MqttEvent.createPublishFail(mBean.getClientKey(), topic, "client is not connected, can not send message", new RuntimeException()));
		}catch (Exception e){
			e.printStackTrace();
			EventBus.getDefault().post(MqttEvent.createPublishFail(mBean.getClientKey(), topic, "publish fail", e));
		}
		return null;
	}

	/**
	 * 订阅主题
	 * @param topic            主题名称数组
	 * @param qos              服务质量数组 0，1，2
	 * @param messageListeners 消息监听器
	 */
	public void subscribe(String[] topic, int[] qos, @Nullable IMqttMessageListener[] messageListeners) {
		Log.d(TAG, "subscribe({" + Arrays.toString(topic) + "}," + Arrays.toString(qos));
		if (mClient == null) {
			EventBus.getDefault().post(MqttEvent.createSubscribeFail(mBean.getClientKey(), topic, "client is null", new NullPointerException()));
			return;
		}
		if (!mClient.isConnected()) {
			EventBus.getDefault().post(MqttEvent.createSubscribeFail(mBean.getClientKey(), topic, "client is disconnect", new IllegalStateException()));
			return;
		}
		IMqttActionListener listener = new SubscribeMqttEventActionListener(mBean.getClientKey(), "subscribe failure", topic);
		try {
			if (messageListeners == null || messageListeners.length == 0) {
				mClient.subscribe(topic, qos, null, listener);
			} else {
				mClient.subscribe(topic, qos, messageListeners);
			}
		} catch (Exception e) {
			e.printStackTrace();
			EventBus.getDefault().post(MqttEvent.createSubscribeFail(mBean.getClientKey(), topic, "subscribe fail", new IllegalStateException()));
		}
	}

	/**
	 * 取消订阅主题
	 * @param topic         主题名称数组
	 */
	void unsubscribe(String[] topic) {
		Log.d(TAG, "unsubscribe({" + Arrays.toString(topic) );

		if (mClient == null) {
			EventBus.getDefault().post(MqttEvent.createUnsubscribeFail(mBean.getClientKey(), topic, "client is null", new NullPointerException()));
			return;
		}
		if (!mClient.isConnected()) {
			EventBus.getDefault().post(MqttEvent.createUnsubscribeFail(mBean.getClientKey(), topic, "client is disconnect", new IllegalStateException()));
			return;
		}
		try {
			mClient.unsubscribe(topic, null, new UnsubscribeMqttEventActionListener(mBean.getClientKey(), "unsubscribe failure", topic));
		} catch (Exception e) {
			e.printStackTrace();
			EventBus.getDefault().post(MqttEvent.createUnsubscribeFail(mBean.getClientKey(), topic, "unsubscribe fail", new IllegalStateException()));
		}
	}

	/**
	 * 获取IMqttDeliveryToken数组
	 */
	public IMqttDeliveryToken[] getPendingDeliveryTokens() {
		if (mClient == null) {
			return null;
		}
		return mClient.getPendingDeliveryTokens();
	}

	/*-------------------------------------- 实现MqttCallbackExtended ----------------------------------------------------*/
	/**
	 * 连接丢失回调
	 * @param why 中断原因
	 *
	 */
	@Override
	public void connectionLost(@Nullable Throwable why) {
		Log.d(TAG, "connectionLost(" + why + ")");
		EventBus.getDefault().post(MqttEvent.createConnectionLost(mBean.getClientKey(), why));
		if (mClient == null){
			return;
		}
		try {
			if(!mBean.getConnectOptions().isAutomaticReconnect()) {
				mClient.disconnect();//没有开启自动重连则手动断开连接
			} else {
				mAlarmPingSender.schedule(100);
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		MqttUtils.releaseWakeLock(mWakeLock);
	}


	@Override
	public void connectComplete(boolean reconnect, String serverURI) {
		EventBus.getDefault().post(MqttEvent.createConnectComplete(mBean.getClientKey(), reconnect, serverURI));
	}

	/**
	 * 发送的消息已到达
	 * @param messageToken 消息令牌
	 */
	@Override
	public void deliveryComplete(IMqttDeliveryToken messageToken) {
		Log.d(TAG, "deliveryComplete(" + messageToken + ")");
		EventBus.getDefault().post(MqttEvent.createDeliveryComplete(mBean.getClientKey(), messageToken));
	}

	/**
	 * 服务端消息到达
	 * @param topic 主题
	 * @param message 消息对象
	 */
	@Override
	public void messageArrived(String topic, MqttMessage message) {
		Log.d(TAG, "messageArrived(" + topic + ",{" + message.toString() + "})");
		String messageId = mMessageStore.saveMessage(mBean.getClientKey(), topic, message);
		DbStoredData data = new DbStoredData(messageId, mBean.getClientKey(), topic,message);
		EventBus.getDefault().post(MqttEvent.createMsgArrived(mBean.getClientKey(), data));
	}
	/*-------------------------------------- 实现MqttCallbackExtended ----------------------------------------------------*/


	/**
	 * 设置客户端离线
	 */
	public void offline() {
		if (mClient == null){
			return;
		}
		if (mClient.isConnected() && !mBean.getConnectOptions().isCleanSession()) {
			connectionLost(new NetworkErrorException("mqtt offline"));
		}
	}
	
	/** 重连 */
	synchronized void reconnect() {
		if (mClient == null) {
			Log.e(TAG, "client is null");
			return;
		}

		if (isConnecting) {
			Log.d(TAG, "the client is connecting");
			return;
		}

		if (!MqttUtils.isOnline(mContext)) {
			Log.d(TAG, "the network is offline, cannot reconnect");
			return;
		}

		if(mBean.getConnectOptions().isAutomaticReconnect()){ //开启自动重连
			Log.i(TAG, "start automatic reconnect");
			try {
				mClient.reconnect();
			} catch (Exception e){
				e.printStackTrace();
				isConnecting = false;
				EventBus.getDefault().post(MqttEvent.createConnectFail(mBean.getClientKey(), "reconnect fail", e));
			}
			return;
		}

		if (!mClient.isConnected() && !mBean.getConnectOptions().isCleanSession()) { // 手动重连
			connect();
		}
	}
	
	/**
	 * 设置断连缓冲配置项
	 * @param bufferOpts 断连缓冲配置项
	 */
	public void setBufferOpts(DisconnectedBufferOptions bufferOpts) {
		this.bufferOpts = bufferOpts;
		if (mClient != null){
			mClient.setBufferOpts(bufferOpts);
		}
	}

	/**
	 * 获取断连缓冲配置项
	 */
	public int getBufferedMessageCount(){
		return mClient != null ? mClient.getBufferedMessageCount() : 0;
	}

	/**
	 * 获取消息
	 * @param bufferIndex 索引
	 */
	public MqttMessage getBufferedMessage(int bufferIndex){
		return mClient != null ? mClient.getBufferedMessage(bufferIndex) : null;
	}

	/**
	 * 删除消息
	 * @param bufferIndex 索引
	 */
	public void deleteBufferedMessage(int bufferIndex){
		if (mClient != null){
			mClient.deleteBufferedMessage(bufferIndex);
		}
	}
}

File: hermes/src/main/java/com/lodz/android/hermes/mqtt/base/MqttConnection.kt
package com.lodz.android.hermes.mqtt.base

import android.accounts.NetworkErrorException
import android.content.Context
import android.os.PowerManager
import com.lodz.android.hermes.modules.HermesLog
import com.lodz.android.hermes.mqtt.base.bean.eun.MqttAction
import com.lodz.android.hermes.mqtt.base.bean.event.MqttEvent
import com.lodz.android.hermes.mqtt.base.bean.source.ClientInfoBean
import com.lodz.android.hermes.mqtt.base.contract.DisconnectMqttEventActionListener
import com.lodz.android.hermes.mqtt.base.contract.PublishMqttEventActionListener
import com.lodz.android.hermes.mqtt.base.contract.SubscribeMqttEventActionListener
import com.lodz.android.hermes.mqtt.base.contract.UnsubscribeMqttEventActionListener
import com.lodz.android.hermes.mqtt.base.db.DbStoredData
import com.lodz.android.hermes.mqtt.base.db.MessageStore
import com.lodz.android.hermes.mqtt.base.db.MessageStoreImpl
import com.lodz.android.hermes.mqtt.base.sender.AlarmPingSender
import com.lodz.android.hermes.utils.IoScope
import com.lodz.android.hermes.utils.HermesUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.eclipse.paho.client.mqttv3.DisconnectedBufferOptions
import org.eclipse.paho.client.mqttv3.IMqttActionListener
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken
import org.eclipse.paho.client.mqttv3.IMqttMessageListener
import org.eclipse.paho.client.mqttv3.IMqttToken
import org.eclipse.paho.client.mqttv3.MqttAsyncClient
import org.eclipse.paho.client.mqttv3.MqttCallbackExtended
import org.eclipse.paho.client.mqttv3.MqttClientPersistence
import org.eclipse.paho.client.mqttv3.MqttMessage
import org.greenrobot.eventbus.EventBus
import java.lang.RuntimeException
import kotlin.concurrent.Volatile

/**
 * MQTT连接对象
 * @author zhouL
 * @date 2023/10/12
 */
class MqttConnection(
    private val mContext: Context, // 上下文
    private val mBean: ClientInfoBean, // 客户端信息
    mPersistence: MqttClientPersistence, // 持久层接口
) : MqttCallbackExtended {

    companion object {
        private const val TAG = "MqttConnection"
    }

    /** 客户端 */
    private var mClient: MqttAsyncClient? = null
    /** 数据库操作方法 */
    private val mMessageStore: MessageStore = MessageStoreImpl(mContext)

    /** 唤醒锁 */
    private val mWakeLock: PowerManager.WakeLock = HermesUtils.getWakeLock(mContext,
        "${javaClass.name} ${mBean.getClientId()} on host ${mBean.getServerURI()}")

    /** ping数据包发送器 */
    private val mAlarmPingSender: AlarmPingSender = AlarmPingSender(mContext)

    /** 连接断开配置 */
    private var mBufferOpts: DisconnectedBufferOptions? = null

    /** 是否连接中 */
    @Volatile
    private var isConnecting = false


    init {
        try {
            mClient = MqttAsyncClient(mBean.getServerURI(), mBean.getClientId(), mPersistence, mAlarmPingSender)
            mClient?.setCallback(this)
        } catch (e: Exception) {
            e.printStackTrace()
            EventBus.getDefault().post(MqttEvent.createFail(mBean.getClientKey(), MqttAction.ACTION_CREATE_CLIENT, "create MqttAsyncClient fail",e))
        }
    }

    /** 连接服务器 */
    fun connect() {
        val client = mClient
        if (client == null) {
            EventBus.getDefault().post(MqttEvent.createConnectFail(mBean.getClientKey(), "client is null",  NullPointerException()))
            return
        }
        HermesLog.d(TAG, "start connect {${mBean.getClientKey()}}")
        if (mBean.getConnectOptions().isCleanSession){// 如果是新的会话则清空数据库缓存
            IoScope().launch { mMessageStore.clearAllMessages(mBean.getClientKey()) }
        }
        if (isConnecting) {
            HermesLog.d(TAG, "now connecting")
            return
        }
        if (client.isConnected) {
            HermesLog.d(TAG, "client is already connected")
            connectSuccess()
            return
        }
        HermesLog.d(TAG, "start connect!")
        isConnecting = true
        try {
            client.connect(mBean.getConnectOptions(), object :IMqttActionListener{
                override fun onSuccess(asyncActionToken: IMqttToken?) {
                    HermesLog.i(TAG, "connect success!")
                    connectSuccess()
                }

                override fun onFailure(asyncActionToken: IMqttToken?, exception: Throwable?) {
                    HermesLog.e(TAG, "connect fail, call connect to reconnect.reason : ${exception?.message}")
                    connectFail("connect failure", exception)
                }
            })
        } catch (e: Exception) {
            e.printStackTrace()
            isConnecting = false
            connectFail("connect fail", e)
        }
        HermesUtils.releaseWakeLock(mWakeLock)
    }

    /** 连接成功 */
    private fun connectSuccess(){
        HermesUtils.acquireWakeLock(mWakeLock)
        EventBus.getDefault().post(MqttEvent.createConnectSuccess(mBean.getClientKey()))
        deliverBacklog()
        isConnecting = false
        HermesUtils.releaseWakeLock(mWakeLock)
    }

    /** 将数据库缓存的到达消息全部返回给用户 */
    private fun deliverBacklog() {
        IoScope().launch {
            val backlog = mMessageStore.getAllMessages(mBean.getClientKey())
            for (data in backlog) {
                launch(Dispatchers.Main) { EventBus.getDefault().post(MqttEvent.createMsgArrived(mBean.getClientKey(), data, mBean.getAckType())) }
            }
        }
    }

    /** 连接失败 */
    private fun connectFail(errorMsg: String, e: Throwable?) {
        HermesUtils.acquireWakeLock(mWakeLock)
        isConnecting = false
        EventBus.getDefault().post(MqttEvent.createConnectFail(mBean.getClientKey(), errorMsg, e ?: RuntimeException()))
        HermesUtils.releaseWakeLock(mWakeLock)
    }

    /** 关闭客户端 */
    fun close(){
        HermesLog.d(TAG, "close client")
        try {
            mClient?.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /** 断开连接，在断开连接之前允许完成现有工作的时间[quiesceTimeout]（以毫秒为单位）。 值为零或更低意味着客户端不会停顿 */
    @JvmOverloads
    fun disconnect(quiesceTimeout: Long = -1) {
        val client = mClient
        if (client == null) {
            EventBus.getDefault().post(MqttEvent.createFail(mBean.getClientKey(), MqttAction.ACTION_DISCONNECT, "client is null",  NullPointerException()))
            return
        }
        HermesLog.d(TAG, "mqtt disconnect {${mBean.getClientKey()}")
        if (!client.isConnected){
            EventBus.getDefault().post(MqttEvent.createDisconnectSuccess(mBean.getClientKey()))
            return
        }
        HermesLog.d(TAG, "start disconnect!")
        val listener = DisconnectMqttEventActionListener(mBean.getClientKey(), "disconnect failure")
        try {
            if (quiesceTimeout > 0) {
                client.disconnect(quiesceTimeout, null, listener)
            } else {
                client.disconnect(null, listener)
            }
        } catch (e: Exception) {
            e.printStackTrace()
            EventBus.getDefault().post(MqttEvent.createDisconnectFail(mBean.getClientKey(), "disconnect fail", e))
        }
        if (mBean.getConnectOptions().isCleanSession){
            IoScope().launch { mMessageStore.clearAllMessages(mBean.getClientKey()) }
        }
        HermesUtils.releaseWakeLock(mWakeLock)
    }

    /** 是否已连接 */
    fun isConnected() = mClient?.isConnected ?: false

    /** 向主题[topic]发送消息[content] */
    fun publish(topic: String, content: String): IMqttDeliveryToken? = publish(topic, content.toByteArray())

    /** 向主题[topic]发送消息，消息内容[payload]，服务质量[qos]一般为0、1、2，MQTT服务器是否保留该消息[isRetained] */
    @JvmOverloads
    fun publish(
        topic: String,
        payload: ByteArray,
        qos: Int = 1,
        isRetained: Boolean = false
    ): IMqttDeliveryToken? {
        val message = MqttMessage(payload)
        message.qos = qos
        message.isRetained = isRetained
        return publish(topic, message)
    }

    /** 向主题[topic]发送消息[message] */
    fun publish(topic: String, message: MqttMessage): IMqttDeliveryToken? {
        val client = mClient
        if (client == null) {
            EventBus.getDefault().post(MqttEvent.createPublishFail(mBean.getClientKey(), topic, "client is null",  NullPointerException()))
            return null
        }
        val listener = PublishMqttEventActionListener(mBean.getClientKey(), "publish failure", topic, message)
        try {
            if (client.isConnected) {
                return client.publish(topic, message, null, listener)
            }
            val opts = mBufferOpts
            if (opts != null && opts.isBufferEnabled) {
                // 虽然客户端未连接，但是缓冲已启用，也允许发送消息
                return client.publish(topic, message, null, listener)
            }
            EventBus.getDefault().post(MqttEvent.createPublishFail(mBean.getClientKey(), topic, "client is not connected, can not send message", RuntimeException()))
        } catch (e: Exception) {
            e.printStackTrace()
            EventBus.getDefault().post(MqttEvent.createPublishFail(mBean.getClientKey(), topic, "publish fail", e))
        }
        return null
    }

    /** 订阅主题[topic]，服务质量[qos]一般为0、1、2，消息监听[listeners] */
    @JvmOverloads
    fun subscribe(topic: String, qos: Int = 1, listeners: Array<IMqttMessageListener>? = null) {
        subscribe(arrayOf(topic), intArrayOf(qos), listeners)
    }

    /** 订阅多主题[topics]，消息监听[listeners] */
    @JvmOverloads
    fun subscribe(topics: Array<String>, listeners: Array<IMqttMessageListener>? = null) {
        val qos = IntArray(topics.size)
        topics.forEachIndexed { i, s ->
            qos[i] = 1
        }
        subscribe(topics, qos, listeners)
    }

    /** 订阅多主题[topics]，服务质量[qos]一般为0、1、2，消息监听[listeners] */
    @JvmOverloads
    fun subscribe(topics: Array<String>, qos: IntArray, listeners: Array<IMqttMessageListener>? = null) {
        HermesLog.d(TAG, "subscribe topics : ${topics.contentToString()} , qos : ${qos.contentToString()}")
        val client = mClient
        if (client == null) {
            EventBus.getDefault().post(MqttEvent.createSubscribeFail(mBean.getClientKey(), topics, "client is null",  NullPointerException()))
            return
        }
        if (!client.isConnected) {
            EventBus.getDefault().post(MqttEvent.createSubscribeFail(mBean.getClientKey(), topics, "client is disconnect",  IllegalStateException()))
            return
        }
        val listener = SubscribeMqttEventActionListener(mBean.getClientKey(), "subscribe failure", topics)
        try {
            if (listeners.isNullOrEmpty()) {
                client.subscribe(topics, qos, null, listener)
            } else {
                client.subscribe(topics, qos, listeners)
            }
        } catch (e: Exception) {
            e.printStackTrace()
            EventBus.getDefault().post(MqttEvent.createSubscribeFail(mBean.getClientKey(), topics, "subscribe fail",  IllegalStateException()))
        }
    }

    /** 取消订阅主题[topic] */
    fun unsubscribe(topic: String) {
        unsubscribe(arrayOf(topic))
    }

    /** 取消订阅主题[topics] */
    fun unsubscribe(topics: Array<String>) {
        HermesLog.d(TAG, "unsubscribe topic : ${topics.contentToString()}")
        val client = mClient
        if (client == null) {
            EventBus.getDefault().post(MqttEvent.createUnsubscribeFail(mBean.getClientKey(), topics, "client is null",  NullPointerException()))
            return
        }
        if (!client.isConnected) {
            EventBus.getDefault().post(MqttEvent.createUnsubscribeFail(mBean.getClientKey(), topics, "client is disconnect",  IllegalStateException()))
            return
        }
        try {
            client.unsubscribe(topics, null,  UnsubscribeMqttEventActionListener(mBean.getClientKey(), "unsubscribe failure", topics))
        } catch (e: Exception) {
            e.printStackTrace()
            EventBus.getDefault().post(MqttEvent.createUnsubscribeFail(mBean.getClientKey(), topics, "unsubscribe fail",  IllegalStateException()))
        }
    }

    /** 获取IMqttDeliveryToken数组 */
    fun getPendingDeliveryTokens() : Array<IMqttDeliveryToken>? = mClient?.pendingDeliveryTokens

    /*-------------------------------------- 实现MqttCallbackExtended ----------------------------------------------------*/

    /** 连接丢失回调，断连原因[cause] */
    override fun connectionLost(cause: Throwable?) {
        HermesLog.d(TAG, "connectionLost case : $cause")
        EventBus.getDefault().post(MqttEvent.createConnectionLost(mBean.getClientKey(), cause))
        val client = mClient ?: return
        try {
            if (!mBean.getConnectOptions().isAutomaticReconnect) {
                client.disconnect()//没有开启自动重连则手动断开连接
            } else {
                mAlarmPingSender.schedule(100)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        HermesUtils.releaseWakeLock(mWakeLock)
    }

    /** 服务端消息到达 */
    override fun messageArrived(topic: String, message: MqttMessage) {
        HermesLog.d(TAG, "messageArrived topic : $topic , message : $message")
        IoScope().launch {
            val messageId = mMessageStore.saveMessage(mBean.getClientKey(), topic, message)
            launch(Dispatchers.Main) {
                EventBus.getDefault().post(MqttEvent.createMsgArrived(mBean.getClientKey(), DbStoredData(messageId, mBean.getClientKey(), topic, message), mBean.getAckType()))
            }
        }
    }

    /** 发送的消息已到达 */
    override fun deliveryComplete(token: IMqttDeliveryToken?) {
        HermesLog.d(TAG, "deliveryComplete IMqttDeliveryToken : $token");
        EventBus.getDefault().post(MqttEvent.createDeliveryComplete(mBean.getClientKey(), token))
    }

    override fun connectComplete(reconnect: Boolean, serverURI: String) {
        isConnecting = false
        EventBus.getDefault().post(MqttEvent.createConnectComplete(mBean.getClientKey(), reconnect, serverURI))
    }

    /*-------------------------------------- 实现MqttCallbackExtended ----------------------------------------------------*/

    /** 设置客户端离线 */
    fun offline() {
        val client = mClient ?: return
        if (client.isConnected && !mBean.getConnectOptions().isCleanSession) {
            connectionLost(NetworkErrorException("mqtt offline"));
        }
    }

    /** 重连 */
    fun reconnect() {
        val client = mClient ?: return
        if (isConnecting) {
            HermesLog.d(TAG, "the client is connecting")
            return
        }
        if (!HermesUtils.isOnline(mContext)) {
            HermesLog.d(TAG, "the network is offline, cannot reconnect")
            return
        }
        if (mBean.getConnectOptions().isAutomaticReconnect) {//开启自动重连
            HermesLog.i(TAG, "start automatic reconnect")
            try {
                isConnecting = true
                client.reconnect()
            } catch (e: Exception) {
                e.printStackTrace()
                isConnecting = false
                EventBus.getDefault().post(MqttEvent.createConnectFail(mBean.getClientKey(), "reconnect fail", e))
            }
            return
        }
        if (!client.isConnected && !mBean.getConnectOptions().isCleanSession) {// 手动重连
            connect()
        }
    }

    /** 设置断连缓冲配置项[bufferOpts] */
    fun setBufferOpts(bufferOpts: DisconnectedBufferOptions) {
        mBufferOpts = bufferOpts
        mClient?.setBufferOpts(bufferOpts)
    }

    /** 获取断连缓冲配置项 */
    fun getBufferedMessageCount(): Int = mClient?.bufferedMessageCount ?: 0

    /** 根据索引[bufferIndex]获取消息 */
    fun getBufferedMessage(bufferIndex: Int): MqttMessage? = mClient?.getBufferedMessage(bufferIndex)

    /** 根据索引[bufferIndex]删除消息 */
    fun deleteBufferedMessage(bufferIndex: Int) {
        mClient?.deleteBufferedMessage(bufferIndex)
    }

    /** 确认消息到达并删除数据库消息编号[messageId]的缓存 */
    fun acknowledgeMessageArrival(messageId: String) {
        IoScope().launch { mMessageStore.deleteArrivedMessage(mBean.getClientKey(), messageId) }
    }
}